lib/codecs/src/encoding/format/syslog.rs

```diff
@@ -0,0 +1,577 @@
+use bytes::{BufMut, BytesMut};
+use tokio_util::codec::Encoder;
+use vector_core::{config::DataType, event::{Event, LogEvent}, schema};
+use chrono::{DateTime, SecondsFormat, SubsecRound, Utc};
+use vrl::value::{ObjectMap, Value};
+use vector_config::configurable_component;
+
+use std::collections::HashMap;
+
+// All of this block is to support the Facility + Severity enums with convenience of string or ordinal config value:
+use std::str::FromStr;
+use strum::{FromRepr, EnumString, VariantNames};
+// `akin` macro for DRY impl to share with both enums due to lack of a `FromRepr` trait:
+use akin::akin;
+// Custom deserialization with serde needed:
+use serde::{Deserialize, Deserializer, de::Error};
+use serde_aux::field_attributes::deserialize_number_from_string;
+use snafu::{Snafu, OptionExt};
+
+/// Config used to build a `SyslogSerializer`.
+#[configurable_component]
+#[derive(Clone, Debug, Default)]
+#[serde(default)]
+pub struct SyslogSerializerConfig {
+    /// Options for the Syslog serializer.
+    pub syslog: SyslogSerializerOptions
+}
+
+impl SyslogSerializerConfig {
+    /// Build the `SyslogSerializer` from this configuration.
+    pub fn build(&self) -> SyslogSerializer {
+        SyslogSerializer::new(&self)
+    }
+
+    /// The data type of events that are accepted by `SyslogSerializer`.
+    pub fn input_type(&self) -> DataType {
+        DataType::Log
+    }
+
+    /// The schema required by the serializer.
+    pub fn schema_requirement(&self) -> schema::Requirement {
+        schema::Requirement::empty()
+    }
+}
+
+/// Syslog serializer options.
+#[configurable_component]
+#[derive(Clone, Debug, Default)]
+// Serde default makes all config keys optional.
+// Each field assigns either a fixed value, or field name (lookup field key to retrieve dynamic value per `LogEvent`).
+#[serde(default)]
+pub struct SyslogSerializerOptions {
+    /// RFC
+    rfc: SyslogRFC,
+    /// Facility
+    #[serde(deserialize_with = "Facility::deserialize")]
+    facility: Facility,
+    /// Severity
+    #[serde(deserialize_with = "Severity::deserialize")]
+    severity: Severity,
+
+    /// App Name
+    app_name: Option<String>,
+    /// Proc ID
+    proc_id: Option<String>,
+    /// Msg ID
+    msg_id: Option<String>,
```

Comment:
> This option is not needed.
> If the message contains a field with a meaning of proc_id, we should use that field. Otherwise, this field should be the PID of Vector.

Quoting the spec:
```   
PROCID is a value that is included in the message, having no
   interoperable meaning, except that a change in the value indicates
   there has been a discontinuity in syslog reporting.  The field does
   not have any specific syntax or semantics; the value is
   implementation-dependent and/or operator-assigned.  The NILVALUE MAY
   be used when no value is provided.
   ```
We use vector in a containerized environment so this being something like say... the podname, or pod uid is more useful then pid... pid is always 1 in the container.  Being able to set this from unique pod information is useful

> Can you remove this option and take the value from the event with a meaning of msgid. We should also update the decoder to insert msgid with this meaning.

Again, is there a reason not to provide a default and make this configurable.   The spec says: 
```
 The MSGID SHOULD identify the type of message.  For example, a
   firewall might use the MSGID "TCPIN" for incoming TCP traffic and the
   MSGID "TCPOUT" for outgoing TCP traffic. Messages with the same
   MSGID should reflect events of the same semantics.  The MSGID itself
   is a string without further semantics.  It is intended for filtering
   messages on a relay or collector.
```
which means to me there may be a possible for a default but it also could be something an admin wants to configure.  We use vector in to kubernetes cluster from various log sources.  I can imagine using this field to distinguish between container and node journal logs